packages/workflow/src/interceptors.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Type definitions and generic helpers for interceptors.
+ *
+ * The Workflow specific interceptors are defined here.
+ *
+ * @module
+ */
+
+import { coresdk } from '@temporalio/proto';
+import { ActivityOptions } from './interfaces';
+import { AnyFunc, OmitLastParam } from './type-helpers';
+
+/** Type of the next function for a given interceptor function */
+export type Next<IF, FN extends keyof IF> = Required<IF>[FN] extends AnyFunc ? OmitLastParam<Required<IF>[FN]> : never;
+
+/** Headers are just a mapping of header name to Payload */
+export type Headers = Map<string, coresdk.common.IPayload>;
+
+/**
+ * Composes all interceptor methods into a single function
+ *
+ * @param interceptors a list of interceptors
+ * @param method the name of the interceptor method to compose
+ * @param next the original function to be executed at the end of the interception chain
+ */
+export function composeInterceptors<I, M extends keyof I>(interceptors: I[], method: M, next: Next<I, M>): Next<I, M> {
+  for (let i = interceptors.length - 1; i >= 0; --i) {
+    const interceptor = interceptors[i];
+    if (interceptor[method] !== undefined) {
+      const prev = next;
+      // We loose type safety here because Typescript can't deduce that interceptor[method] is of type Next<I, M>
+      next = ((input: any) => (interceptor[method] as any)(input, prev)) as any;
```

Comment:
Doesn't look like there's a way for an interceptor to report an error, like throwing a validation error if activity params are invalid for example. Is that something worth supporting?

Reply:
You can throw an error and fail the interception